x11/xauth.go

```diff
@@ -0,0 +1,157 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package x11
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+)
+
+// See https://cgit.freedesktop.org/xorg/lib/libXau/tree/AuRead.c and
+// https://cgit.freedesktop.org/xorg/lib/libXau/tree/include/X11/Xauth.h
+// for details about the actual file format.
+type xauth struct {
+	Family  uint16
+	Address []byte
+	Number  []byte
+	Name    []byte
+	Data    []byte
+}
+
+func readChunk(f *os.File) ([]byte, error) {
+	b := [2]byte{}
+	n, err := f.Read(b[:])
+	if err != nil {
+		return nil, err
+	} else if n != 2 {
+		return nil, fmt.Errorf("Could not read enough bytes")
+	}
+
+	size := int(binary.BigEndian.Uint16(b[:]))
+	chunk := make([]byte, size)
+	n, err = f.Read(chunk)
+	if err != nil {
+		return nil, err
+	} else if n != size {
+		return nil, fmt.Errorf("Could not read enough bytes")
+	}
+
+	return chunk, nil
+}
+
+func (xa *xauth) readFromFile(f *os.File) error {
+	b := [2]byte{}
+	n, err := f.Read(b[:])
+	if err != nil {
+		return err
+	} else if n != 2 {
+		return fmt.Errorf("Could not read enough bytes")
+	}
```

Comment:
These 7 lines of code are duplicated here at the top of readChunk and extremely similar in the next part in readChunk. I wonder if it would make sense to abstract this out into `readBytes(n int)`, then you call readBytes(2) to get the Family, readBytes(2) to get the size then readBytes(size) to get the chunk.

Also, you reference the X sources (which is great!) but it would be good to explicitly comment that Family is two bytes, and len for the strings is two bytes.